File: leetcode/572-subtree-of-another-tree/main.py
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


"""
    1st approach: brute force checking
    - for each node, check it equals to target tree

    Time    O(n^2)
    Space   O(h)
    232ms beats 54.85%
    24jan2019
"""


class Solution(object):
    def isSubtree(self, s, t):
        """
        :type s: TreeNode
        :type t: TreeNode
        :rtype: bool
        """
        if s is None and t is None:
            return True
        elif s is None or t is None:
            return False

        if s.val == t.val:
            if self.isIdentical(s, t) == True:
                return True

        left = self.isSubtree(s.left, t)
        right = self.isSubtree(s.right, t)
        return left or right

    def isIdentical(self, a, b):
        if a is None and b is None:
            return True
        elif a is None or b is None:
            return False
        mid = a.val == b.val
        left = self.isIdentical(a.left, b.left)
        right = self.isIdentical(a.right, b.right)
        return mid and left and right


"""
    2nd approach:
    - serialize both trees
    - check if the 2nd serialized representation is within the 1st serialized representation

    corner case: [12] and [2]
    - so we have to wrap the whole tree within a pair of parentheses

    Time    O(S+T)
    Space   O(h)
    68 ms, faster than 96.06%
"""


class Solution(object):
    def isSubtree(self, s, t):
        """
        :type s: TreeNode
        :type t: TreeNode
        :rtype: bool
        """
        sStr = self.serialize(s)
        tStr = self.serialize(t)
        if tStr in sStr:
            return True
        return False

    def serialize(self, node):
        if node == None:
            return '()'
        left = self.serialize(node.left)
        right = self.serialize(node.right)
        return '('+str(node.val) + left + right + ')'

# helpers


def printTree(node):
    if node == None:
        return
    printTree(node.left)
    print(node.val)
    printTree(node.right)


a = TreeNode(1)
b = TreeNode(2)
c = TreeNode(3)
d = TreeNode(4)
e = TreeNode(5)
f = TreeNode(6)
a.left = b
a.right = c
b.left = d
b.right = e
c.left = f
s = Solution().serialize(a)
print(s)
ds = Solution().deserialize(s)
printTree(ds)

File: leetcode/572-subtree-of-another-tree/main.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

/*
	Questions to ask:
	- will there be empty nodes s or t ?
	- if both empty, what should i return?
	- so what i need to check is whether the trees have the same structure and values, but not the addresses/pointers of the nodes?
	amazon OA: https://wdxtub.com/interview/14520850399861.html
*/

/*
	1st approach
	1. dfs to look for the node having the same value with target root
	2. if yes, check the left and right subtrees are identical to the target
	Time	O(n^2) for each node, check its children
	Space	O(h)
	24ms beats 100%
	24jan2019
*/
func isSubtree(s *TreeNode, t *TreeNode) bool {
	if s == nil && t == nil {
		return true
	} else if s == nil || t == nil {
		return false
	}

	if s.Val == t.Val {
		if isIdentical(s, t) == true {
			return true
		}
	}
	left := isSubtree(s.Left, t)
	right := isSubtree(s.Right, t)
	return left || right
}

func isIdentical(a *TreeNode, b *TreeNode) bool {
	if a == nil && b == nil {
		return true
	} else if a == nil || b == nil {
		return false
	}
	mid := a.Val == b.Val
	left := isIdentical(a.Left, b.Left)
	right := isIdentical(a.Right, b.Right)
	return mid && left && right
}

func main() {

}
